### PR TITLE
[FLINK-13938][yarn] Enable configure shared libraries on YARN

### PR DESCRIPTION
**Pull request by TisonKun**
## What is the purpose of the change

Enable configure shared libraries on YARN. See also https://issues.apache.org/jira/browse/FLINK-13938

## Brief change log

- Add command-line option `-ysl` `--yarnsharedLibs` and config option `yarn.shared-libraries`
- respect options

## Verifying this change

I am looking for adding a test case but it seems hard without a real HDFS env. Maybe e2e tests required. But one can still verify it manually.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes, enable configure shared libs on YARN)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

cc @wangyang0918

### CODE REVIEW

**Comment by TisonKun** [comment]
cc @yanghua @MalcolmSanders

File: docs/_includes/generated/yarn_config_configuration.html [hunk 1]
```diff
@@ -102,6 +102,11 @@
             <td style="word-wrap: break-word;">(none)</td>
             <td>When a Flink job is submitted to YARN, the JobManager’s host and the number of available processing slots is written into a properties file, so that the Flink client is able to pick those details up. This configuration parameter allows changing the default location of that file (for example for environments sharing a Flink installation between users).</td>
         </tr>
+        <tr>
+            <td><h5>yarn.shared-libraries</h5></td>
+            <td style="word-wrap: break-word;">(none)</td>
+            <td>specify the path to use public visibility feature of YARN NodeManager localizing resources.</td>
```
Comment (yanghua) [inline docs/_includes/generated/yarn_config_configuration.html]: Shall we capitalize the first letter? Like other config option description.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java [hunk 2]
```diff
@@ -135,6 +138,8 @@
 	/** Lazily initialized list of files to ship. */
 	private final List<File> shipFiles = new LinkedList<>();
 
+	private final List<Path> sharedLibs = new LinkedList<>();
```
Comment (yanghua) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java]: If we want to test this feature like ship files, we may expose `getter/setter` methods of this field.

File: flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java [hunk 3]
```diff
@@ -194,6 +195,7 @@ public FlinkYarnSessionCli(
 		applicationId = new Option(shortPrefix + "id", longPrefix + "applicationId", true, "Attach to running YARN session");
 		queue = new Option(shortPrefix + "qu", longPrefix + "queue", true, "Specify YARN queue.");
 		shipPath = new Option(shortPrefix + "t", longPrefix + "ship", true, "Ship files in the specified directory (t for transfer)");
+		sharedLibs = new Option(shortPrefix + "sl", longPrefix + "sharedLibs", true, "Upload a copy of flink libs beforehand and specify the path to use public visibility feature of YARN NodeManager localizing resources.");
```
Comment (yanghua) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: Can we add a test case about this config option or the whole feature like ship files?

File: flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java [hunk 4]
```diff
@@ -215,6 +215,14 @@
 				.noDefaultValue()
 				.withDescription("A semicolon-separated list of directories to be shipped to the YARN cluster.");
 
+	public static final ConfigOption<List<String>> SHARED_LIBRARIES =
+			key("yarn.shared-libraries")
+				.stringType()
+				.asList()
+				.noDefaultValue()
+				.withDescription("specify the path to use public visibility feature of " +
```
Comment (yanghua) [inline flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java]: Also, shall we capitalize the first letter? Like other config option description.

**Comment by TisonKun** [comment]
I'd like to add a unit test but fail to have an idea how to start a mini HDFS. Let me think of how to introduce it within YarnTestBase...

**Comment by TisonKun** [comment]
I've manually verify the changes locally.

My rough thought about testing is using a test job running with YarnTestBase and verify the resource, maybe we have to verify no files uploaded by Flink also.

File: flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java (hunk 4 above)
Comment (TisonKun) [reply flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java]: Make sense. Will update.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java [hunk 5]
```diff
@@ -1237,6 +1262,47 @@ public FileVisitResult visitFile(java.nio.file.Path file, BasicFileAttributes at
...
+		for (Path sharedLibPath : sharedLibs) {
+			if (fs.isFile(sharedLibPath)) {
+				String classpath = setupSharedLib(sharedLibPath, fs, remotePaths, localResources, envShipFileList);
+				classPaths.add(classpath);
+			} else { // directory
+				RemoteIterator<LocatedFileStatus> i = fs.listFiles(sharedLibPath, true);
+				while (i.hasNext()) {
+					LocatedFileStatus fileStatus = i.next();
+					if (fileStatus.isFile()) {
+						String classpath = setupSharedLib(fileStatus.getPath(), fs, remotePaths, localResources, envShipFileList);
```
Comment (TisonKun) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java]: Here we flatten all files in the root relative path of program. Maybe it is better we keep the directory layout. Locally changed and verifying...

**Comment by TisonKun** [comment]
closed for now in favor of another issue.

After discussion with @wangyang0918 I notice FLINK-13938 is not what proposed here.

**Comment by walterddr** [comment]
sorry for the late follow up @TisonKun . I come to understand that this PR is for something different from FLINK-13938. did we create a new JIRA ticket? I think the approach this PR implements is highly valuable especially for a production cluster setup :-) --Rong

**Pull request closed without merging**